spark/src/main/java/org/apache/iceberg/spark/source/SparkPlannerUtil.java

```diff
@@ -0,0 +1,231 @@
+          snapshot.snapshotId(), snapshot.timestampMillis(), scan.filter());
+
+      Listeners.notifyAll(
+          new ScanEvent(scan.table().toString(), snapshot.snapshotId(), scan.filter(), scan.schema()));
+
+      return planFilesFor(spark, scan, snapshot);
+    } else {
+      LOG.info("Scanning empty table {}", scan.table());
+      return CloseableIterable.empty();
+    }
+  }
+
+  private static CloseableIterable<FileScanTask> planFilesFor(SparkSession spark, TableScan scan, Snapshot snapshot) {
+    //This is different for different BaseTableScan implementations
+    if (scan instanceof DataTableScan) {
+      return planDataTableScanFiles(spark, (DataTableScan) scan, snapshot);
+    } else {
+      throw new UnsupportedOperationException(String.format("Unable to do distributed planning for %s",
+          scan.getClass()));
+    }
+  }
+
+  //TODO these two functions probably belong in util now
+  private static String metadataTableName(Table table, MetadataTableType type) {
+    return metadataTableName(table, table.toString(), type);
+  }
+
+  private static String metadataTableName(Table table, String tableName, MetadataTableType type) {
+    if (tableName.contains("/")) {
+      return tableName + "#" + type;
+    } else if (tableName.startsWith("hadoop.")) {
+      // for HadoopCatalog tables, use the table location to load the metadata table
+      // because IcebergCatalog uses HiveCatalog when the table is identified by name
+      return table.location() + "#" + type;
+    } else if (tableName.startsWith("hive.")) {
+      // HiveCatalog prepend a logical name which we need to drop for Spark 2.4
+      return tableName.replaceFirst("hive\\.", "") + "." + type;
+    } else {
+      return tableName + "." + type;
+    }
+  }
+
+  private static CloseableIterable<FileScanTask> planDataTableScanFiles(
+      SparkSession spark,
+      DataTableScan scan,
+      Snapshot snapshot) {
+
+    //TODO Currently this approach reads all manifests, no manifest filtering - Maybe through pushdowns or table read
+    // options?
+    //Create a dataframe of all DataFile entries
+    String dataFilesMetadataTable = metadataTableName(scan.table(), MetadataTableType.FILES);
+    Dataset<Row> dataFileRows =
+        spark.read()
+            .format("iceberg")
+            .load(dataFilesMetadataTable);
+
+    Types.StructType partitionStruct = DataFile.getType(scan.table().spec().partitionType());
+    StructType sparkSchema = dataFileRows.schema();
+
+    Dataset<SparkDataFile> dataFiles =
+        dataFileRows.mapPartitions((Iterator<Row> it) -> {
+          //Wrap is mutable here and not thread safe
+              SparkDataFile container = new SparkDataFile(partitionStruct, sparkSchema);
+              return Streams.stream(it).map(container::wrap).iterator();
+              },
+            Encoders.javaSerialization(SparkDataFile.class));
```

Comment:
As an alternative to using java serialization and making `SparkDataFile` serializable, we could work with `Dataset<Row>` directly and do `mapPartitions` and instantiate `SparkDataFile` and evaluators on executors, wrap Spark rows into data files, filter them, unwrap matching to rows. Then we could collect this to the driver and define a lazy transformation to `FileScanTask`.

Reply:
I've been thinking about this more and I think combining all of our logic into a single mapper makes sense

Reply:
+1 to working with Spark native primitives (aka `Dataset<Row>`) and `mapPartitions`. To me, the biggest benefit is that the visualization of the tasks in the DAG in the Spark UI / History Server would be much more identifiable and there's nothing more frustrating to my users than stages that have very little information about them. Though this should not run nearly as long as the ones that people complain about (in the worst case, there's one job that has a step that basically blocks the history server UI and makes it seem like the job has stalled for 2 or 3 days 😅  - but the job sure is resilient and I blame Spark's data source v1 which didn't provide much support UI wise).

I also agree that combining the logic into a single mapper would likely make sense and would greatly reduce SerDe / serializability concerns.

Reply:
While I was able to remove the serialization here, we still need SparkDataFiles to be serializable since we need to send them back out in the ScanTask objects. Please see the updated commits, lots of new changes